### PR TITLE
feat!: improve error on malformed `abi`

### PR DESCRIPTION
**Pull request by hal3e**
closes: https://github.com/FuelLabs/fuels-rs/issues/1343

BREAKING CHANGE: 
- `AbigenTarget` now holds the parsed `FullProgramAbi` and the path.
- fields of `AbigenTarget` and `Abi` are `pub(crate)`
- `Source` is removed in favor of `Abi`

### Checklist
- [x] I have linked to any relevant issues.
- [x] I have added tests that prove my fix is effective or that my feature works.
- [x] I have added necessary labels.
- [x] I have done my best to ensure that my PR adheres to [the Fuel Labs Code Review Standards](https://github.com/FuelLabs/rfcs/blob/master/text/code-standards/external-contributors.md).
- [x] I have requested a review from the relevant team or maintainers.

### CODE REVIEW

**Review by segfault-magnet — DISMISSED** [review]
Would be nice if we could catch this before `Abigen`. Sometime while we still have the `Span` so that we may underline the ABI part of the macro.

We can change the `AbigenTarget` to contain the `FullAbiProgram` instead of the source. 

But maybe for another time. Maybe an issue if you agree? Or a point in the neverending wish list @digorithm ?

**Comment by hal3e** [comment]
I will give it a try int this PR

File: packages/fuels-code-gen/src/program_bindings/abigen/abigen_target.rs [hunk 1]
```diff
@@ -3,49 +3,19 @@ use std::{convert::TryFrom, path::PathBuf, str::FromStr};
...
-    let source = Source::parse(abi_source).expect("failed to parse JSON ABI");
-
-    let json_abi_str = source.get().expect("failed to parse JSON ABI from string");
-    let abi = FullProgramABI::from_json_abi(&json_abi_str)?;
-    let path = source.path();
-    Ok(Abi { path, abi })
+#[derive(Debug, Clone)]
+pub struct Abi {
+    pub path: Option<PathBuf>,
+    pub abi: FullProgramABI,
```
[the remark this answers is not in the document]
Comment (hal3e) [reply packages/fuels-code-gen/src/program_bindings/abigen/abigen_target.rs:18]: Good catch. I would solve this by adding a method to `Abi` called `parse` that sets the right values and remove the pub access. Similarly, `AbiTarget` gets `new` and we remove `pub` access.

**Review by segfault-magnet — DISMISSED** [review]
Ty for the effort :1st_place_medal:

File: packages/fuels-code-gen/src/program_bindings/abigen.rs [hunk 2]
```diff
@@ -83,7 +77,7 @@ impl Abigen {
     }
 
     fn generate_all_bindings(
-        parsed_targets: Vec<ParsedAbigenTarget>,
+        parsed_targets: Vec<AbigenTarget>,
```
Comment (MujkicA) [inline packages/fuels-code-gen/src/program_bindings/abigen.rs:80]:
```suggestion
        targets: Vec<AbigenTarget>,
```

**Review by MujkicA — APPROVED** [review]
Nice 🚀

**Review by MujkicA — APPROVED** [review]

**Review by segfault-magnet — APPROVED** [review]

**Review by Salka1988 — APPROVED** [review]

**Review by digorithm — APPROVED** [review]
:shipit:

**Pull request merged**